src/exo/LoopIR_scheduling.py

```diff
@@ -1493,6 +1504,194 @@ def map_e(self, e):
         return None
 
 
+def get_reads_of_stmts(stmts):
+    reads = []
+    for s in stmts:
+        if isinstance(s, LoopIR.Assign):
+            reads += get_reads_of_expr(s.rhs)
+            reads += sum([get_reads_of_expr(idx) for idx in s.idx], [])
+        elif isinstance(s, LoopIR.Reduce):
+            reads += get_reads_of_expr(s.rhs)
+            reads += sum([get_reads_of_expr(idx) for idx in s.idx], [])
+        elif isinstance(s, LoopIR.If):
+            reads += get_reads_of_stmts(s.body)
+            if s.orelse:
+                reads += get_reads_of_stmts(s.orelse)
+        elif isinstance(s, LoopIR.Seq):
+            reads += get_reads_of_stmts(s.body)
+        elif isinstance(s, LoopIR.Call):
+            reads += sum([get_reads_of_expr(arg) for arg in s.args], [])
+        elif isinstance(s, (LoopIR.WindowStmt, LoopIR.WriteConfig)):
+            raise NotImplementedError("WindowStmt and WriteConfig not supported yet")
+        elif isinstance(s, (LoopIR.Pass, LoopIR.Alloc, LoopIR.Free)):
+            pass
+        else:
+            raise NotImplementedError(f"unknown stmt type {type(s)}")
+    return reads
+
+
+def get_writes_of_stmts(stmts):
+    writes = []
+    for s in stmts:
+        if isinstance(s, LoopIR.Assign):
+            writes += [(s.name, s.type)]
+        elif isinstance(s, LoopIR.Reduce):
+            writes += [(s.name, s.type)]
+        elif isinstance(s, LoopIR.If):
+            writes += get_writes_of_stmts(s.body)
+            if s.orelse:
+                writes += get_writes_of_stmts(s.orelse)
+        elif isinstance(s, LoopIR.Seq):
+            writes += get_writes_of_stmts(s.body)
+        elif isinstance(s, (LoopIR.WindowStmt, LoopIR.WriteConfig)):
+            raise NotImplementedError("WindowStmt and WriteConfig not supported yet")
+        elif isinstance(s, (LoopIR.Pass, LoopIR.Alloc, LoopIR.Free, LoopIR.Call)):
+            pass
+        else:
+            raise NotImplementedError(f"unknown stmt type {type(s)}")
+    return writes
+
+
+def DoLiftConstant(assign_cursor, loop_cursor):
+    orig_proc = assign_cursor.get_root()
+    assign_s = assign_cursor._node
+    loop = loop_cursor._node
+    constant = None
+    constant_src_stmt = None
+
+    for (name, typ) in get_reads_of_stmts(loop.body):
+        if assign_s.name == name and assign_s.type == typ:
+            raise SchedulingError(
+                "cannot lift constant because the buffer is read in the loop body"
+            )
+
+    def check_only_scaled_reduces(stmts):
```

Comment:
This is way too deeply nested for a local function. It should be rewritten to either return:

1. `None` upon failure, or
2. `constant` upon success

The `constant_src_stmt` variable is only used inside this function and doesn't need to be `nonlocal`.

It's also quite complex with nested ifs that drop-off in an effort to do nothing. This sort of thing needs comments or to be restructured.

Reply:
I restructured the code significantly, hopefully its cleaner now.